domain/application/service/application.go

```diff
@@ -1541,11 +1546,39 @@ func (s *Service) GetApplicationStatus(ctx context.Context, appID coreapplicatio
 		return nil, internalerrors.Errorf("application ID: %w", err)
 	}
 
-	info, err := s.st.GetApplicationStatus(ctx, appID)
+	applicationStatus, err := s.st.GetApplicationStatus(ctx, appID)
+	if err != nil {
+		return nil, internalerrors.Capture(err)
+	}
+	if applicationStatus.Status != application.WorkloadStatusUnset {
+		return decodeWorkloadStatus(applicationStatus)
+	}
+
+	// The application status is unset. However, we can still derive the status
+	// of the application using the workload statuses of all the application's
+	// units.
+	//
+	// NOTE: It is possible that between these two calls to state someone else
+	// calls SetApplicationStatus and changes the status. This would potentially
+	// lead to an out of date status being returned here. In this specific case,
+	// we don't mind so long as we have 'eventual' (i.e. milliseconds) consistency.
+
+	unitStatuses, err := s.st.GetUnitWorkloadStatusesForApplication(ctx, appID)
 	if err != nil {
 		return nil, internalerrors.Capture(err)
 	}
-	return decodeWorkloadStatus(info)
+
+	deocodedStatuses := []corestatus.StatusInfo{}
+	for _, unitStatus := range unitStatuses {
+		status, err := decodeWorkloadStatus(&unitStatus)
+		if err != nil {
+			return nil, internalerrors.Capture(err)
+		}
+		deocodedStatuses = append(deocodedStatuses, *status)
+	}
+	dervicedStatus := corestatus.DeriveStatus(deocodedStatuses)
```

Comment:
I'm fine with this for the moment, but I think long term we want the DeriveStatus to work on the domain/application types, not the core status ones.